src/main.cpp

```diff
@@ -3496,20 +3496,15 @@ bool CheckBlock(const CBlock &block, CValidationState &state, bool fCheckPOW, bo
     return true;
 }
 
-bool CheckIndexAgainstCheckpoint(const CBlockIndex *pindexPrev,
-    CValidationState &state,
-    const CChainParams &chainparams,
-    const uint256 &hash)
+bool CheckAgainstCheckpoint(unsigned int height, const uint256 &hash, const CChainParams &chainparams)
 {
-    if (*pindexPrev->phashBlock == chainparams.GetConsensus().hashGenesisBlock)
-        return true;
-
-    int nHeight = pindexPrev->nHeight + 1;
-    // Don't accept any forks from the main chain prior to last checkpoint
-    CBlockIndex *pcheckpoint = Checkpoints::GetLastCheckpoint(chainparams.Checkpoints());
-    if (pcheckpoint && nHeight < pcheckpoint->nHeight)
-        return state.DoS(100, error("%s: forked chain older than last checkpoint (height %d)", __func__, nHeight));
```

Comment:
if we start banning again then we may have trouble on testnet...but perhaps we could just issue an fDisconnect=true ?

Reply:
what about banning only on mainnet?

Reply:
I think that we should differentiate testnet from mainnet as little as possible.

This is a good observation, I hadn't noticed it.

But I tend to dislike banning because it increases the potential for a netsplit, and does not soak up the attacker's efforts, leaving him to connect to someone else.  And I especially dislike a helper function making that decision, but only if "state" is properly threaded back up the stack and then acted on (in 2 of of 3 cases it is not).

We do need to ban for egregious stuff, but in this case, we've probably requested this header or block so now we ban the node because it gave them to us?  Regardless, we won't be asking for it again so there is no harm in remaining connected.

Reply:
I somewhat share your concern about net split and this is way I think we should do what @ptschip proposed, i.e. using `fDisconnect=true` and let our reconnect logic to take care of the problematic peers w/o banning them on the spot.